src/gtest/test_sideChain.cpp

```diff
@@ -0,0 +1,1196 @@
+    //Insert Sc
+    uint256 newScId = uint256S("a1b2");
+    CAmount initialAmount = 1000;
+    int scCreationHeight = 7;
+    aTransaction = createSideChainTxWith(newScId, initialAmount);
+    coinViewCache.UpdateScInfo(aTransaction, aBlock, scCreationHeight);
+
+    int coinMaturityHeight = scCreationHeight + Params().ScCoinsMaturity();
+    int lookupBlockHeight = coinMaturityHeight;
+
+    //prerequisites
+    ASSERT_TRUE(coinViewCache.sidechainExists(newScId))<<"Test requires existing sc";
+    ASSERT_TRUE(lookupBlockHeight == coinMaturityHeight)
+        <<"Test requires attempting to mature coins at maturity height";
+
+    //test
+    bool res = coinViewCache.ApplyMatureBalances(lookupBlockHeight, aBlockUndo);
+
+    //checks
+    EXPECT_TRUE(res);
+    EXPECT_TRUE(coinViewCache.getScInfoMap().at(newScId).balance == initialAmount)
+        <<"Current balance is "<<coinViewCache.getScInfoMap().at(newScId).balance
+        <<" expected one is "<<initialAmount;
+}
+
+TEST_F(SideChainTestSuite, CoinsInScCreationDoNotModifyScBalanceAfterCoinMaturity) {
+    //Insert Sc
+    uint256 newScId = uint256S("a1b2");
+    CAmount initialAmount = 1000;
+    int scCreationHeight = 11;
+    aTransaction = createSideChainTxWith(newScId, initialAmount);
+    coinViewCache.UpdateScInfo(aTransaction, aBlock, scCreationHeight);
+
+    int coinMaturityHeight = anHeight + Params().ScCoinsMaturity();
+    int lookupBlockHeight = coinMaturityHeight + 1;
+
+    //prerequisites
+    ASSERT_TRUE(coinViewCache.sidechainExists(newScId))<<"Test requires existing sc";
+    ASSERT_TRUE(lookupBlockHeight > coinMaturityHeight)
+        <<"Test requires attempting to mature coins after their maturity height";
+
+    //test
+    bool res = coinViewCache.ApplyMatureBalances(lookupBlockHeight, aBlockUndo);
+
+    //check
+    EXPECT_FALSE(res);
+    EXPECT_TRUE(coinViewCache.getScInfoMap().at(newScId).balance < initialAmount)
+        <<"Current balance is "<<coinViewCache.getScInfoMap().at(newScId).balance
+        <<" while initial amount is "<<initialAmount;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+/////////////////////////// RestoreImmatureBalances ///////////////////////////
+///////////////////////////////////////////////////////////////////////////////
+TEST_F(SideChainTestSuite, RestoringFromUndoBlockAffectBalance) {
+    //insert a sidechain
+    uint256 newScId = uint256S("ca1985");
+    CAmount initialAmount = 34;
+    int scCreationHeight = 71;
+    aTransaction = createSideChainTxWith(newScId, initialAmount);
+    coinViewCache.UpdateScInfo(aTransaction, aBlock, scCreationHeight);
+
+    //let balance mature
+    int maturityHeight = scCreationHeight + Params().ScCoinsMaturity();
+    coinViewCache.ApplyMatureBalances(maturityHeight, aBlockUndo);
+    CAmount scBalance = coinViewCache.getScInfoMap().at(newScId).balance;
+
+    int amountToUndo = 17;
```

Comment:
CAmount should be used instead of int (even if it is typedef to int64_t)

Reply:
Agree. Fixed changing type to CAmount.